src/java/org/apache/cassandra/cache/ChunkCache.java

```diff
@@ -125,6 +131,39 @@ public boolean equals(Object obj)
         }
     }
 
+    static class BufferFuture
+    {
+        private final Key key;
+        private final Buffer buffer;
+        private boolean loaded;
+
+        public BufferFuture(Key key, Buffer buffer)
+        {
+            this.key = key;
+            this.buffer = buffer;
+        }
+
+        int size()
+        {
+            return key.file.chunkSize();
+        }
+
+        synchronized void release()
+        {
+            buffer.release();
+        }
+
+        public synchronized Buffer loadBufferIfNeeded()
+        {
+            if (!loaded && buffer.references.get() > 0)
```

Comment:
Maybe add a guard to protect from calling it after `release` ?

Reply:
Could this reference be some other thread's (after `release` is already called), freed while we are reading the chunk?

Reply:
also "release" is tricky, because basically we can prevent loading only if the refcount goes to zero, as you can release and reference it multiple times